maas/resource_maas_boot_source.go

```diff
@@ -1,12 +1,150 @@
 package maas
 
 import (
+	"context"
 	"fmt"
 
 	"github.com/canonical/gomaasclient/client"
 	"github.com/canonical/gomaasclient/entity"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 )
 
+const defaultURL = "http://images.maas.io/ephemeral-v3/stable/"
+const defaultKeyring = "/snap/maas/current/usr/share/keyrings/ubuntu-cloudimage-keyring.gpg"
+
+func resourceMAASBootSource() *schema.Resource {
+	return &schema.Resource{
+		Description:   "Provides a resource to manage the MAAS boot source.",
+		CreateContext: resourceBootSourceCreate,
+		ReadContext:   resourceBootSourceRead,
+		UpdateContext: resourceBootSourceUpdate,
+		DeleteContext: resourceBootSourceDelete,
+
+		Schema: map[string]*schema.Schema{
+			"created": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The creation time of the boot source.",
+			},
+			"keyring_data": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Description:   "The data on the keyring for the boot source.",
+				ConflictsWith: []string{"keyring_filename"},
```

Comment:
It seems that we should use ExactlyOneOf instead of ConflictsWith. Based on MAAS codebase, at least one of them must be set: https://github.com/canonical/maas/blob/master/src/maasserver/models/bootsource.py#L33-L48

```suggestion
				ExactlyOneOf: []string{"keyring_filename"},
```